### PR TITLE
windows: Add file dialog using IFileOpenDialog

### PR DESCRIPTION
**Pull request by the-jasoney**
Release Notes:

- Added a file dialog for Windows

### CODE REVIEW

**Comment by the-jasoney** [comment]
@cla-bot check

File: crates/gpui/src/platform/windows/platform.rs [hunk 1]
```diff
@@ -290,17 +298,91 @@ impl Platform for WindowsPlatform {
 
     // todo!("windows")
     fn open_url(&self, url: &str) {
-        // todo!("windows")
+        unimplemented!()
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/platform.rs:301]: lets leave this as a comment.

File: crates/gpui/src/platform/windows/platform.rs [hunk 2]
```diff
@@ -290,17 +298,91 @@ impl Platform for WindowsPlatform {
...
     // todo!("windows")
     fn on_open_urls(&self, callback: Box<dyn FnMut(Vec<String>)>) {
         self.inner.callbacks.lock().open_urls = Some(callback);
     }
 
-    // todo!("windows")
+    #[allow(overflowing_literals)]
     fn prompt_for_paths(&self, options: PathPromptOptions) -> Receiver<Option<Vec<PathBuf>>> {
-        unimplemented!()
+        println!("prompt_for_paths {:?}", options);
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/platform.rs:311]: Debugging prints should be removed.

File: crates/gpui/src/platform/windows/platform.rs [hunk 3]
```diff
@@ -290,17 +298,91 @@ impl Platform for WindowsPlatform {
...
+                    return;
+                }
+            }
+
+            let mut results = unsafe { folder_dialog.GetResults().unwrap() };
+
+            let mut paths: Vec<PathBuf> = Vec::new();
+            for i in 0..unsafe { results.GetCount().unwrap() } {
+                let mut item: IShellItem = unsafe { results.GetItemAt(i).unwrap() };
+                let mut path = unsafe { item.GetDisplayName(SIGDN(0x80058000)).unwrap() }; // SIGDN_FILESYSPATH
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/platform.rs:363]: I think these constants should be replaced theirs windows crate equivalents, e.g. https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/UI/Shell/constant.SIGDN_FILESYSPATH.html

File: crates/gpui/src/platform/windows/platform.rs [hunk 4]
```diff
@@ -290,17 +298,91 @@ impl Platform for WindowsPlatform {
...
+
+                paths.push(PathBuf::from(unsafe { path.to_string().unwrap() }));
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/platform.rs:365]: I don't think this is safe on windows? IIRC windows UTF-16 and rust strings are UTF-8. At least, this should be `to_string_lossy()`

File: crates/gpui/src/platform/windows/platform.rs [hunk 5]
```diff
@@ -290,17 +298,91 @@ impl Platform for WindowsPlatform {
...
+        unimplemented!()
     }
 
     // todo!("windows")
     fn on_open_urls(&self, callback: Box<dyn FnMut(Vec<String>)>) {
         self.inner.callbacks.lock().open_urls = Some(callback);
     }
 
-    // todo!("windows")
+    #[allow(overflowing_literals)]
```
Comment (mikayla-maki) [inline crates/gpui/src/platform/windows/platform.rs:309]: This should be removed.

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

**Comment by mikayla-maki** [comment]
Could you include a screenshot of the working file dialog as well?

File: crates/gpui/src/platform/windows/platform.rs (hunk 3 above)
Comment (the-jasoney) [reply crates/gpui/src/platform/windows/platform.rs]: Replaced all of them except `HRESULT(0x800704C7)`, wasn't able to find a constant for it

**Comment by the-jasoney** [comment]
Screenshot 2024-03-06 124312

File: crates/gpui/src/platform/windows/platform.rs (hunk 1 above)
Comment (the-jasoney) [reply crates/gpui/src/platform/windows/platform.rs]: resolved with 751a2e1

**Comment by mikayla-maki** [comment]
I notice that [this part of another PR](https://github.com/zed-industries/zed/pull/8959/files#diff-5b5e706f04c15d77efd23989ec5b1b1cf73d36144e2d066a074de165533ecaeaR139) initializes COM for the length of the app, through a different API. Could we use that approach instead? Or is there a compelling reason to keep the approach you're using?

**Comment by mikayla-maki** [comment]
Once the COM initialization is figured out, and the merge conflicts resolved, I think I'd prefer to merge this one as it looks a little nicer.

**Comment by the-jasoney** [comment]
> I notice that [this part of another PR](https://github.com/zed-industries/zed/pull/8959/files#diff-5b5e706f04c15d77efd23989ec5b1b1cf73d36144e2d066a074de165533ecaeaR139) initializes COM for the length of the app, through a different API. Could we use that approach instead? Or is there a compelling reason to keep the approach you're using?

I'm pretty sure the approach in the PR would work.

File: crates/gpui/src/platform/windows/platform.rs [hunk 6]
```diff
@@ -298,9 +300,78 @@ impl Platform for WindowsPlatform {
...
+            }
+
+            let hr = unsafe { folder_dialog.Show(None) };
+
+            if hr.is_err() {
+                if hr.unwrap_err().code() == HRESULT(0x800704C7u32 as i32)  { // user canceled error
+                    if let Some(tx) = tx.take() {
+                        tx.send(None).unwrap();
+                    }
+                    return;
```
Comment (kazatsuyu) [inline crates/gpui/src/platform/windows/platform.rs:343]: `CoUninitialize` is not called here.

**Comment by mikayla-maki** [comment]
@the-jasoney when the merge conflicts are resolved, I'd be down to merge this :)

**Comment by JunkuiZhang** [comment]
> Once the COM initialization is figured out, and the merge conflicts resolved, I think I'd prefer to merge this one as it looks a little nicer.

The `CoInitialize` stuff should be deleted, by [Microsoft](https://learn.microsoft.com/en-us/windows/win32/api/ole2/nf-ole2-oleinitialize):
> `OleInitialize` calls `CoInitializeEx` internally to initialize the COM library on the current apartment.

And since the COM library is initialized on an apartment only once, `CoInitialize` using in this PR actually does nothing. See Microsoft's comments:
> Typically, the COM library is initialized on an apartment only once. Subsequent calls will succeed, as long as they do not attempt to change the concurrency model of the apartment, but will return S_FALSE.

**Comment by mikayla-maki** [comment]
Now that https://github.com/zed-industries/zed/pull/8959 has been merged, we should get this branch up to date with main and remove the COM initialization / tear down. Then we can merge :)

**Comment by mikayla-maki** [comment]
Seems this needs another merge conflict resolution, and a `cargo fmt` before we can merge it :)

**Review by mikayla-maki — APPROVED** [review]

**Pull request merged**